### PR TITLE
Validate driver configuration

### PR DESCRIPTION
**Pull request by iverberk**
This PR validates driver configuration option in the jobspec. Each driver maintains a schema which the options must adhere to. This includes checking for required fields. The type checking is done strictly and only for the top-level type. Currently it doesn't recurse into arrays or maps to do deeper inspection of types. Nomad will tell you which options have errors and prevent the job from getting scheduled.

The test-fixtures were automatically formatted by vim-hclfmt (hence the big diff) but I also had fix numerous configuration errors in the fixtures, which the validation pointed out quite nicely :-)

### CODE REVIEW

File: client/driver/docker.go [hunk 1]
```diff
@@ -126,6 +127,73 @@ func NewDockerDriver(ctx *DriverContext) Driver {
 	return &DockerDriver{DriverContext: *ctx}
 }
 
+// Validate is used to validate the driver configuration
+func (d *DockerDriver) Validate(config map[string]interface{}) error {
+	fd := &fields.FieldData{
+		Raw: config,
+		Schema: map[string]*fields.FieldSchema{
```
Comment (diptanu) [inline client/driver/docker.go]: missing the option `intercative`

Comment (iverberk) [inline client/driver/docker.go]: ah, yes, it was added recently and I missed it. Fixed.

File: helper/fields/data.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
+package fields
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/mitchellh/mapstructure"
+)
+
+// FielData contains the raw data and the schema that the data should adhere to
```
Comment (dadgar) [inline helper/fields/data.go]: Small typo "FieldData"

File: helper/fields/data.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+type FieldData struct {
+	Raw    map[string]interface{}
+	Schema map[string]*FieldSchema
+}
+
+// Cycle through raw data and validate conversions in the schema. Also check
```
Comment (dadgar) [inline helper/fields/data.go]: Start function comments with the name of the function. `Validate cycles through....`

File: helper/fields/data.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+	switch schema.Type {
+	case TypeBool, TypeInt, TypeMap, TypeArray, TypeString:
+		return d.getPrimitive(k, schema)
+	default:
+		return nil, false,
+			fmt.Errorf("unknown field type %s for field %s", schema.Type, k)
+	}
+}
+
+func (d *FieldData) getPrimitive(
```
Comment (dadgar) [inline helper/fields/data.go]: Can you put a comment on this method

File: jobspec/parse.go [hunk 5]
```diff
@@ -537,6 +538,22 @@ func parseTasks(jobName string, taskGroupName string, result *[]*structs.Task, l
 					return err
 				}
 			}
+
+			// Instantiate a driver to validate the configuration
+			d, err := driver.NewDriver(
```
Comment (dadgar) [inline jobspec/parse.go]: I would also add this to the task.Validate() method in structs.go

Comment (iverberk) [inline jobspec/parse.go]: This will cause an import cycle (driver imports structs and vice-versa). Do you see any way around that?

Comment (dadgar) [inline jobspec/parse.go]: You can create a new package so that structs can depend on that which in turn depends on driver but driver wont depend on it

**Pull request merged**